---
title: 'Comparison Between Experiments'
output:
  html_document:
    code_folding: 'hide'
---

```{r load_packages, echo=TRUE, include=FALSE}
library(tidyverse) ; library(reshape2) ; library(glue) ; library(plotly) ; library(RColorBrewer)
library(viridis) ; require(gridExtra) ; library(GGally) ; library(ggExtra) ; library(ggpubr)
library(VennDiagram) ; library(venneuler)
library(biomaRt)
library(expss)
library(knitr)
library(anRichment) ; library(BrainDiseaseCollection)
```

```{r load_datasets, message=FALSE, warning=FALSE}

###################################################################################################
# Gene Expression data

# All Regions
load('./../../AllRegions/Data/preprocessed_data.RData')
datExpr_all = datExpr %>% data.frame
DE_info_all = DE_info %>% data.frame %>% mutate(ID = rownames(.))
datMeta_full = datMeta %>% dplyr::select(-matches('SV|Picard'))

# Frontal
load('./../../Frontal/Data/preprocessed_data.RData')
datExpr_frontal = datExpr %>% data.frame
DE_info_frontal = DE_info %>% data.frame %>% mutate(ID = rownames(.))
datMeta = datMeta %>% dplyr::select(-matches('SV|Picard'))
datMeta_full = rbind(datMeta_full, datMeta) %>% distinct(Sample_ID, .keep_all = TRUE)

# Occipital
load('./../../Occipital/Data/preprocessed_data.RData')
datExpr_occipital = datExpr %>% data.frame
DE_info_occipital = DE_info %>% data.frame %>% mutate(ID = rownames(.))
datMeta = datMeta %>% dplyr::select(-matches('SV|Picard'))
datMeta_full = rbind(datMeta_full, datMeta) %>% distinct(Sample_ID, .keep_all = TRUE)

# Without Occipital
load('./../../NoOccipital/Data/preprocessed_data.RData')
datExpr_no_occipital = datExpr %>% data.frame
DE_info_no_occipital = DE_info %>% data.frame %>% mutate(ID = rownames(.))
datMeta = datMeta %>% dplyr::select(-matches('SV|Picard'))
datMeta_full = rbind(datMeta_full, datMeta) %>% distinct(Sample_ID, .keep_all = TRUE)

# Get Entrez Gene ID for all genes
all_genes = unique(c(DE_info_all$ID, DE_info_frontal$ID,
                     DE_info_occipital$ID, DE_info_no_occipital$ID))
getinfo = c('ensembl_gene_id','entrezgene')
mart = useMart(biomart='ENSEMBL_MART_ENSEMBL', dataset='hsapiens_gene_ensembl', host='feb2014.archive.ensembl.org')
genes_EntrezGeneID = getBM(attributes=getinfo, filters=c('ensembl_gene_id'), values=all_genes, mart=mart)
colnames(genes_EntrezGeneID)[1] = 'ID'

###################################################################################################

# SFARI Genes
SFARI_genes = read_csv('./../../../SFARI/Data/SFARI_genes_08-29-2019_w_ensembl_IDs.csv')
SFARI_genes = SFARI_genes[!duplicated(SFARI_genes$ID) & !is.na(SFARI_genes$ID),]

###################################################################################################

# GO Neuronal annotations: regex 'neuron' in GO functional annotations and label the genes that make a match as neuronal
GO_annotations = read.csv('./../../AllRegions/Data/genes_GO_annotations.csv')
GO_neuronal = GO_annotations %>% filter(grepl('neuron', go_term)) %>% 
              mutate('ID'=as.character(ensembl_gene_id)) %>% 
              dplyr::select(-ensembl_gene_id) %>% distinct(ID) %>%
              mutate('Neuronal'=1)

###################################################################################################

# Prepare datasets for enrichment analysis
GO_col = buildGOcollection(organism = 'human', verbose = 0)
internal_col = internalCollection(organism = 'human')
MillerAIBS_col = MillerAIBSCollection(organism = 'human')
BrainDisease_col = BrainDiseaseCollection(organism = 'human')
combined_col = mergeCollections(GO_col, internal_col, MillerAIBS_col, BrainDisease_col)



rm(GO_annotations, datMeta, datExpr, datGenes, dds, DE_info, GO_col, internal_col, MillerAIBS_col,
   BrainDisease_col, getinfo, mart)
```

## Pool of Genes

The genes studied in each experiment were very similar
```{r}
all_genes = unique(genes_EntrezGeneID$ID)

genes_df = data.frame('All' = all_genes %in% DE_info_all$ID,
                      'Fro' = all_genes %in% DE_info_frontal$ID,
                      'Occ' = all_genes %in% DE_info_occipital$ID,
                      'NoO' = all_genes %in% DE_info_no_occipital$ID)

cat(paste0('Total number of genes: ', nrow(genes_df)))
```


```{r, fig.width=5, fig.height=5}
grid.newpage()
grid.draw(draw.quad.venn(sum(genes_df$All), sum(genes_df$Fro), sum(genes_df$Occ), sum(genes_df$NoO),
          sum(genes_df$All*genes_df$Fro), sum(genes_df$All*genes_df$Occ), sum(genes_df$All*genes_df$NoO),
          sum(genes_df$Fro*genes_df$Occ), sum(genes_df$Fro*genes_df$NoO), sum(genes_df$Occ*genes_df$NoO),
          sum(genes_df$All*genes_df$Fro*genes_df$Occ), sum(genes_df$All*genes_df$Fro*genes_df$NoO),
          sum(genes_df$All*genes_df$Occ*genes_df$NoO), sum(genes_df$Fro*genes_df$Occ*genes_df$NoO),
          sum(genes_df$All*genes_df$Fro*genes_df$Occ*genes_df$NoO),
          category = c('AllRegions', 'Frontal', 'Occipital', 'WoOccipital'), 
          fill = hcl(h=seq(15, 375, length=5), l=65, c=100)[-5], fontfamily = rep('sans-serif',15),
          alpha = rep(0.25,4), lty = rep('blank', 4), cat.fontfamily = rep('sans-serif',4)))
```

## DE Genes

***

### Analyse each experiment independently

```{r}
genes_df = data.frame('All' = all_genes %in% DE_info_all$ID[DE_info_all$padj<0.05],
                      'Fro' = all_genes %in% DE_info_frontal$ID[DE_info_frontal$padj<0.05],
                      'Occ' = all_genes %in% DE_info_occipital$ID[DE_info_occipital$padj<0.05],
                      'NoO' = all_genes %in% DE_info_no_occipital$ID[DE_info_no_occipital$padj<0.05],
                      'ID' = all_genes)
rownames(genes_df) = all_genes
genes_df = genes_df[genes_df$All+genes_df$Fro+genes_df$Occ+genes_df$NoO>0,]

cat(paste0('Total number of DE genes: ', nrow(genes_df)))

plot(venneuler(combinations = c('All Regions'=sum(genes_df$All), 'Frontal'=sum(genes_df$Fro),
                                'Occipital'=sum(genes_df$Occ), 'Without Occipital'=sum(genes_df$NoO))))
```

#### 1. All Regions

```{r, EA_all_regions}
perform_enrichment = function(DE_info){

  EA_dataset = DE_info %>% left_join(genes_EntrezGeneID, by = 'ID')
  
  # Perform Enrichment Analysis
  enrichment = enrichmentAnalysis(classLabels = EA_dataset$DE, identifiers = EA_dataset$entrezgene,
                                  refCollection = combined_col, #useBackground = 'given', 
                                  threshold = 1e-4, thresholdType = 'Bonferroni',
                                  getOverlapEntrez = FALSE, getOverlapSymbols = TRUE)
  
  enrichment_df = enrichment$enrichmentTable %>% filter(class==1) %>% 
                  dplyr::select(dataSetID, shortDataSetName, inGroups, Bonferroni, FDR, enrichmentRatio,
                                effectiveClassSize, effectiveSetSize, nCommonGenes) %>%
                  filter(Bonferroni<0.05) %>% arrange(Bonferroni, desc(enrichmentRatio))
  
  return(enrichment_df)
}

enrichment_input = genes_df %>% mutate(DE = as.numeric(All == 1))
enrichment_df = perform_enrichment(enrichment_input)
kable(enrichment_df, caption = 'Enriched terms for dataset including All Brain Regions')
```

<br>

#### 2. Frontal

```{r, EA_Frontal}
enrichment_input = genes_df %>% mutate(DE = as.numeric(Fro == 1))
enrichment_df = perform_enrichment(enrichment_input)
kable(enrichment_df, caption = 'Enriched terms for the Frontal Lobe dataset')
```

<br>

#### 3. Occipital

```{r, EA_Occipital}
enrichment_input = genes_df %>% mutate(DE = as.numeric(Occ == 1))
enrichment_df = perform_enrichment(enrichment_input)
kable(enrichment_df, caption = 'Enriched terms for the Occipital Lobe dataset')
```

<br>

#### 4. Without Occipital

```{r, EA_wo_Occipital}
enrichment_input = genes_df %>% mutate(DE = as.numeric(NoO == 1))
enrichment_df = perform_enrichment(enrichment_input)
kable(enrichment_df, caption = 'Enriched terms for dataset excluding the Occipital Lobe')
```

<br><br>

***

### Analyse interactions between datasets

```{r}
plot(venneuler(genes_df[, colnames(genes_df) != 'ID']))
```

```{r, fig.width=5, fig.height=5}
grid.newpage()
grid.draw(draw.quad.venn(sum(genes_df$All), sum(genes_df$Fro), sum(genes_df$Occ), sum(genes_df$NoO),
          sum(genes_df$All*genes_df$Fro), sum(genes_df$All*genes_df$Occ), sum(genes_df$All*genes_df$NoO),
          sum(genes_df$Fro*genes_df$Occ), sum(genes_df$Fro*genes_df$NoO), sum(genes_df$Occ*genes_df$NoO),
          sum(genes_df$All*genes_df$Fro*genes_df$Occ), sum(genes_df$All*genes_df$Fro*genes_df$NoO),
          sum(genes_df$All*genes_df$Occ*genes_df$NoO), sum(genes_df$Fro*genes_df$Occ*genes_df$NoO),
          sum(genes_df$All*genes_df$Fro*genes_df$Occ*genes_df$NoO),
          category = c('AllRegions', 'Frontal', 'Occipital', 'WoOccipital'), 
          fill = hcl(h=seq(15, 375, length=5), l=65, c=100)[-5], fontfamily = rep('sans-serif',15),
          alpha = rep(0.25,4), lty = rep('blank', 4), cat.fontfamily = rep('sans-serif',4)))
```

```{r}
getinfo = c('ensembl_gene_id','external_gene_id')
mart = useMart(biomart='ENSEMBL_MART_ENSEMBL', dataset='hsapiens_gene_ensembl', host='feb2014.archive.ensembl.org')
common_DE_genes = getBM(attributes=getinfo, filters=c('ensembl_gene_id'), 
                        values=rownames(genes_df)[genes_df$All*genes_df$Fro*genes_df$Occ*genes_df$NoO>0],
                        mart=mart)

common_DE_genes = common_DE_genes %>% dplyr::rename('gene_symbol' = 'external_gene_id') %>%
                  left_join(SFARI_genes, by=c('ensembl_gene_id'='ID')) %>%
                  dplyr::select(ensembl_gene_id, gene_symbol, `gene-score`, syndromic)

kable(common_DE_genes, caption = 'Genes found to be DE in all four experiments')
```

***

### Analyse interactions between datasets without Frontal Lobe experiment
```{r, fig.width=5, fig.height=5}
grid.newpage()
grid.draw(draw.triple.venn(sum(genes_df$All), sum(genes_df$NoO), sum(genes_df$Occ),
                           sum(genes_df$All*genes_df$NoO), sum(genes_df$Occ*genes_df$NoO), sum(genes_df$All*genes_df$Occ),
                           sum(genes_df$All*genes_df$NoO*genes_df$Occ),
          category = c('AllRegions', 'Without Occipital', 'Occipital'), 
          fill = hcl(h=seq(15, 375, length=4), l=65, c=100)[-4], fontfamily = rep('sans-serif',7),
          alpha = rep(0.25,3), lty = rep('blank', 3), cat.fontfamily = rep('sans-serif',3)))
```
<br>

#### 1. DE in All Regions dataset only

These genes are DE only when analysing the whole dataset (not when studying the Occipital Lobe, nor when removing it and studying the rest), so they need information from all areas together to be significant enough

```{r 1_EA_all_regions_only}
enrichment_input = genes_df %>% filter(All) %>% mutate(DE = as.numeric(NoO + Occ == 0))
enrichment_df = perform_enrichment(enrichment_input)
kable(enrichment_df, caption = 'Enriched terms only in the All Regions dataset')
```

<br>

#### 2. DE in Occipital Lobe dataset only

These genes are DE only when analysing the Occipital Lobe alone, so they are not enriched in the other regions and the signal is not strong enough to survive when adding other brain regions

```{r 2_EA_Occipital_only}
enrichment_input = genes_df %>% filter(Occ) %>% mutate(DE = as.numeric(All + NoO == 0))
enrichment_df = perform_enrichment(enrichment_input)
kable(enrichment_df, caption = 'Enriched terms only for the dataset studying the Occipital Lobe')
```

<br>

#### 3. DE in dataset excluding Occipital Lobe only

These genes are not DE in the occipital lobe, and when including the occipital lobe in the samples, the signal is not strong enough to still be identified as statistically significant, only when we remove the occipital lobe it becomes significant

```{r 3_EA_wo_Occipital_only}
enrichment_input = genes_df %>% filter(NoO) %>% mutate(DE = as.numeric(All + Occ == 0))
enrichment_df = perform_enrichment(enrichment_input)
kable(enrichment_df, caption = 'Enriched terms only for the dataset excluding the Occipital Lobe')
```

***

#### 4. DE in All Regions and Occipital samples, but not when removing the Occipital samples

These genes are DE in the occipital lobe and their signal is strong enough to still be identified as significant when combining them with the samples from the rest of the brain regions

```{r 4_EA_all_regions_and_Occipital}
enrichment_input = genes_df %>% filter(All|Occ) %>% mutate(DE = as.numeric(All*Occ == 1 & NoO == 0))
enrichment_df = perform_enrichment(enrichment_input)
kable(enrichment_df, caption = 'Enriched terms for the dataset studying the Occipital Lobe and when studying all the brain regions together, but not when excluding the Occipital samples')
```

<br>

#### 5. DE in All Regions and excluding the Occipital samples, but not when isolating the Occipital samples

These genes are not differentially expressed in the Occipital lobe, but they are in another brain region, but the signal is strong enough to survive even when we include all the regions in the analysis

```{r 5_EA_all_regions_and_wo_Occipital}
enrichment_input = genes_df %>% filter(All|NoO) %>% mutate(DE = as.numeric(All*NoO == 1 & Occ == 0))
enrichment_df = perform_enrichment(enrichment_input)
kable(enrichment_df, caption = 'Enriched terms for the dataset studying the Occipital Lobe and when studying all the brain regions together, but not when excluding the Occipital samples')
```

<br>

#### 6. DE when studying the Occipital lobe as well as when excluding it from the analysis, but not DE when considering all the brain regions

These genes are DE in the Occipital lobe, as well as in another brain region, but perhaps the direction of the differential expression is opposite in both brain regions, which eliminates the effect when studying them together? It's a weird case, so it makes sense that not many genes fall into this category (77)

```{r 6_EA_Occipital_and_wo_Occipital}
enrichment_input = genes_df %>% filter(Occ|NoO) %>% mutate(DE = as.numeric(Occ*NoO == 1 & All == 0))
enrichment_df = perform_enrichment(enrichment_input)
kable(enrichment_df, caption = 'Enriched terms for the dataset studying the Occipital lobe as well as when excluding it from the analysis, but not DE when considering all the brain regions')
```

***

#### 7. DE in all three experiments

These genes are DE both in the Occipital lobe and in another brain region, and the signals are strong enough for them to be recognisable separately as well as when combining all the brain regions together

```{r 7_EA_all_three}
enrichment_input = genes_df %>% filter(All|Occ|NoO) %>% mutate(DE = as.numeric(All*Occ*NoO == 1))
enrichment_df = perform_enrichment(enrichment_input)
kable(enrichment_df, caption = 'Enriched terms for the dataset studying the Occipital lobe as well as when excluding it from the analysis, but not DE when considering all the brain regions')
```
