---
title: 'SFARI Genes'
output:
  html_document:
    code_folding: 'hide'
---

```{r load_packages, echo=TRUE, warning=FALSE, message=FALSE}
#setwd('/afs/inf.ed.ac.uk/user/s17/s1725186/Documents/PhD-Models/FirstPUModel/RMarkdowns')

library(tidyverse) ; library(reshape2) ; library(glue) ; library(plotly) ; library(plotlyutils)
library(RColorBrewer) ; library(viridis) ; require(gridExtra) ; library(GGally)
library(Rtsne)
library(ClusterR)
library(DESeq2)
library(expss)
library(knitr)
```

Load preprocessed dataset (preprocessing code in 19_10_14_data_preprocessing.Rmd)
```{r load_dataset, message=FALSE, warning=FALSE}
# Gandal dataset
load('./../Data/preprocessed_data.RData')
datExpr = datExpr %>% data.frame
DE_info = DE_info %>% data.frame

# GO Neuronal annotations: regex 'neuron' in GO functional annotations and label the genes that make a match as neuronal
GO_annotations = read.csv('./../Data/genes_GO_annotations.csv')
GO_neuronal = GO_annotations %>% filter(grepl('neuron', go_term)) %>% 
              mutate('ID'=as.character(ensembl_gene_id)) %>% 
              dplyr::select(-ensembl_gene_id) %>% distinct(ID) %>%
              mutate('Neuronal'=1)

# SFARI Genes
SFARI_genes = read_csv('./../../../SFARI/Data/SFARI_genes_08-29-2019_w_ensembl_IDs.csv')
SFARI_genes = SFARI_genes[!duplicated(SFARI_genes$ID) & !is.na(SFARI_genes$ID),]

# Update DE_info with SFARI and Neuronal information
DE_info = DE_info %>% mutate('ID'=rownames(.)) %>% left_join(SFARI_genes, by='ID') %>% 
  mutate(`gene-score`=ifelse(is.na(`gene-score`), 'Others', `gene-score`)) %>%
  distinct(ID, .keep_all = TRUE) %>% left_join(GO_neuronal, by='ID') %>%
  mutate(Neuronal=ifelse(is.na(Neuronal), 0, Neuronal)) %>%
  mutate(gene.score=ifelse(`gene-score`=='Others' & Neuronal==1, 'Neuronal', `gene-score`), significant=padj<0.05 & !is.na(padj)) %>%
  mutate(Group = factor(ifelse(gene.score %in% c('Neuronal','Others'), gene.score, 'SFARI'), levels = c('SFARI', 'Neuronal', 'Others')))


SFARI_colour_hue = function(r) {
  pal = c('#FF7631','#FFB100','#E8E328','#8CC83F','#62CCA6','#59B9C9','#b3b3b3','#808080','gray','#d9d9d9')[r]
}

rm(GO_annotations)
```

--- 

### SFARI Gene list
```{r}
cat(paste0('There are ', length(unique(SFARI_genes$`gene-symbol`)), ' genes with a SFARI score'))
```

There are 979 genes with a SFARI score. but to map them to gene expression mapa we had to map the gene names to their corresponding ensembl IDs
<br><br>

#### Mapping SFARI Gene names to Ensembl IDs

```{r}
cat(paste0('There are ', nrow(SFARI_genes), ' Ensembl IDs corresponding to the ',
             length(unique(SFARI_genes$`gene-symbol`)),' genes in the SFARI Gene dataset'))
```
<br><br>

- Since a gene can have more than one ensembl ID, there were some one-to-many mappings between a gene name and ensembl IDs, so that's why we ended up with 1090 rows in the SFARI_genes dataset.

- The details about how the genes were annotated with their Ensembl IDs can be found in 20_02_06_get_ensembl_ids.html

```{r}
cat(paste0('There are ', sum(is.na(SFARI_genes$`gene-score`)) ,
             ' genes in the SFARI list without a score, of which ',
             sum(is.na(SFARI_genes$`gene-score`) & SFARI_genes$syndromic==0),
             ' don\'t have syndromic tag either (Why include them then???)'))
```

--- 

### Exploratory Analysis


```{r}
cat(paste0('There are ', sum(SFARI_genes$ID %in% rownames(datExpr)), ' SFARI Genes in the expression dataset (~',
             round(100*mean(SFARI_genes$ID %in% rownames(datExpr))),'%)'))

cat(paste0('Of these, only ', sum(DE_info$`gene-score`!='Others'), ' have an assigned score'))
```

**From now on, we're only going to focus on the 903 genes with a score**

Gene count by SFARI score:
```{r}
table_info = DE_info %>% apply_labels(`gene-score` = 'SFARI Gene Score', syndromic = 'Syndromic Tag',
                                      Neuronal = 'Neuronal Function', gene.score = 'Gene Score') %>%
             mutate(syndromic = as.logical(syndromic), Neuronal = as.logical(Neuronal))

cro(table_info$`gene-score`)
```

Gene count by Syndromic tag:
```{r}
cro(table_info$syndromic)
```

<br>
GO Neuronal annotations:
```{r}
cat(glue(sum(GO_neuronal$ID %in% rownames(datExpr)), ' genes have neuronal-related annotations'))

cat(glue(sum(SFARI_genes$ID %in% GO_neuronal$ID),' of these genes have a SFARI score'))

cro(table_info$gene.score[DE_info$`gene-score` %in% c('1','2','3','4','5','6')],
    list(table_info$Neuronal[DE_info$`gene-score` %in% c('1','2','3','4','5','6')], total()))
```
<br>

***

## All SFARI scores together

<br>

### Gene Expression

<br>

Larger mean expression than the other two groups, smaller SD than Neuronal genes

```{r, warning=FALSE, fig.width=10}
plot_data = data.frame('ID'=rownames(datExpr), 'MeanExpr'=rowMeans(datExpr), 'SDExpr'=apply(datExpr,1,sd)) %>% 
            left_join(DE_info, by='ID') %>% mutate(Group = factor(ifelse(gene.score %in% c('Neuronal','Others'), gene.score, 'SFARI'), 
                                                                  levels = c('SFARI', 'Neuronal', 'Others')))

p1 = ggplotly(plot_data %>% ggplot(aes(Group, MeanExpr, fill=Group)) + geom_boxplot() + 
              scale_fill_manual(values=c('#00A4F7', SFARI_colour_hue(r=c(8,7)))) + theme_minimal() +
              theme(legend.position='none'))

p2 = ggplotly(plot_data %>% ggplot(aes(Group, SDExpr, fill=Group)) + geom_boxplot() + 
              scale_fill_manual(values=c('#00A4F7', SFARI_colour_hue(r=c(8,7)))) + theme_minimal() +
              ggtitle('Mean Expression (left) and SD (right) comparison between SFARI Genes and the rest of the dataset') + 
              theme(legend.position='none'))

subplot(p1, p2, nrows=1)


rm(plot_data, p1, p2)
```

### Log Fold Change


Proportion of over- and under-expressed genes is very similar between groups: approximately half

```{r, warning=FALSE, message=FALSE}
DE_info %>% mutate(direction = ifelse(log2FoldChange>0, 'over-expressed', 'under-expressed')) %>% group_by(Group, direction) %>% tally(name = 'over_expr') %>% 
            filter(direction == 'over-expressed') %>% ungroup %>% left_join(DE_info %>% group_by(Group) %>% tally(name = 'Total'), by = 'Group') %>% ungroup %>%
            mutate('prop_over_expr' = round(over_expr/Total,3)) %>% dplyr::select(-direction) %>% kable
```

Smaller lFC magnitude than Neuronal genes and similar but slightly lower than the rest of the genes

```{r}
ggplotly(DE_info %>% ggplot(aes(x=Group, y=abs(log2FoldChange), fill=Group)) + geom_boxplot() + 
         scale_fill_manual(values=c('#00A4F7', SFARI_colour_hue(r=c(8,7)))) + ylab('logFoldChange Magnitude') + xlab('Group') + theme_minimal() +
         theme(legend.position='none'))
```

- SFARI Genes, as a group, have less genes with high (positive) lFC than the rest of the genes in the dataset

- Perhaps the opposite is true for the genes with the highest (negative) lFC, although this pattern is weaker

```{r, warning=FALSE, message=FALSE}
plot_data = DE_info  %>% dplyr::select(Group, log2FoldChange) %>%
            mutate(quant = cut(log2FoldChange, breaks = quantile(log2FoldChange, probs = seq(0,1,0.05)) %>% as.vector, labels = FALSE),
                   value_range = cut(log2FoldChange, breaks = quantile(log2FoldChange, probs = seq(0,1,0.05)) %>% as.vector)) %>% 
            filter(Group == 'SFARI') %>% group_by(quant, value_range) %>% tally %>% ungroup %>%
            left_join(DE_info  %>% dplyr::select(Group, log2FoldChange) %>%
                      mutate(quant = cut(log2FoldChange, breaks = quantile(log2FoldChange, probs = seq(0,1,0.05)) %>% as.vector, labels = FALSE)) %>% 
                      group_by(quant) %>% tally(name = 'tot') %>% ungroup) %>% mutate(p = 100*n/tot)

ggplotly(plot_data %>% ggplot(aes(quant, p)) + geom_bar(stat = 'identity', fill = '#00A4F7', aes(id = value_range)) + geom_smooth(color = 'gray', alpha = 0.1) + 
         geom_hline(yintercept = 100*mean(DE_info$Group == 'SFARI'), color = 'gray', linetype = 'dotted') +
         xlab('Log Fold Change Quantiles') + ylab('% of SFARI Genes') + ggtitle('Percentage of SFARI Genes for lFC Quantiles') + theme_minimal())

cat('LFC values for each quantile:')

quants = data.frame('Quantile' = 1:20, 'LFC Range' = cut(DE_info$log2FoldChange, breaks = quantile(DE_info$log2FoldChange, probs = seq(0,1,0.05)) %>% as.vector) %>%
         table %>% names)

quants %>% kable

rm(quants)
```
<br>


### DEA

Lower proportion of DE genes than genes with Neuronal annotation but higher than the rest of the genes

```{r, warning=FALSE, message=FALSE}
DE_info %>% group_by(Group, significant) %>% tally(name = 'DE') %>% filter(significant) %>% ungroup %>%
            left_join(DE_info %>% group_by(Group) %>% tally(name = 'Total'), by = 'Group') %>% ungroup %>% mutate('prop_DE' = round(DE/Total,2)) %>% 
            dplyr::select(-significant) %>% kable
```

SFARI Genes have consistently a lower percentage of DE genes than the Genes with Neuronal annotations but slightly higher than the rest of the genes

```{r warning=FALSE, message=FALSE, fig.width=10}

lfc_list = seq(1, 1.2, 0.01)

all_counts = data.frame('group'='All', 'n'=as.character(nrow(DE_info)))
Others_counts = data.frame('group'='Others', n=as.character(sum(DE_info$Group == 'Others')))
Neuronal_counts = data.frame('group'='Neuronal', n=as.character(sum(DE_info$Neuronal)))
lfc_counts_all = DE_info %>% filter(Group == 'SFARI') %>% tally %>%
                 mutate('group'='SFARI', 'n'=as.character(n)) %>%
                 dplyr::select(group, n) %>%
                 bind_rows(Neuronal_counts, Others_counts, all_counts) %>%
                 mutate('lfc'=-1) %>%  dplyr::select(lfc, group, n)

for(lfc in lfc_list){
  
  # Recalculate DE_info with the new threshold (p-values change)
  DE_genes = results(dds, lfcThreshold=log2(lfc), altHypothesis='greaterAbs') %>% data.frame %>% mutate('ID'=rownames(.)) %>% 
             left_join(DE_info %>% dplyr::select(ID, Neuronal, gene.score, Group), by = 'ID') %>% filter(padj<0.05 & abs(log2FoldChange)>log2(lfc))

  
  # Calculate counts by groups
  all_counts = data.frame('group'='All', 'n'=as.character(nrow(DE_genes)))
  Others_counts = data.frame('group'='Others', n=as.character(sum(DE_genes$Group == 'Others')))
  Neuronal_counts = data.frame('group'='Neuronal', n=as.character(sum(DE_genes$Neuronal)))
  lfc_counts = DE_genes %>% filter(Group == 'SFARI') %>% tally %>%
               mutate('group'='SFARI', 'n'=as.character(n)) %>%
               bind_rows(Neuronal_counts, Others_counts, all_counts) %>%
               mutate('lfc'=lfc) %>% dplyr::select(lfc, group, n)
  
  
  # Update lfc_counts_all
  lfc_counts_all = lfc_counts_all %>% bind_rows(lfc_counts)
}

# Add missing entries with 0s
lfc_counts_all = expand.grid('group'=unique(lfc_counts_all$group), 'lfc'=unique(lfc_counts_all$lfc)) %>% 
  left_join(lfc_counts_all, by=c('group','lfc')) %>% replace(is.na(.), 0)

# Calculate percentage of each group remaining
tot_counts = DE_info %>% filter(Group == 'SFARI') %>% tally() %>%
             mutate('group'='SFARI', 'tot'=n) %>% dplyr::select(group, tot) %>%
             bind_rows(data.frame('group'='Neuronal', 'tot'=sum(DE_info$Neuronal)),
                       data.frame('group' = 'Others', 'tot' = sum(DE_info$Group == 'Others')),
                       data.frame('group'='All', 'tot'=nrow(DE_info)))

lfc_counts_all = lfc_counts_all %>% filter(lfc!=-1) %>% #, group!='Others') %>% 
                 left_join(tot_counts, by='group') %>% mutate('perc'=round(100*as.numeric(n)/tot,2))


# Plot change of number of genes
ggplotly(lfc_counts_all %>% filter(group != 'All') %>% mutate(group = factor(group, levels = c('SFARI', 'Neuronal', 'Others'))) %>%
         ggplot(aes(lfc, perc, color=group)) + geom_point(aes(id=n)) + geom_line() +
         scale_color_manual(values=c('#00A4F7', SFARI_colour_hue(r=c(8,7)))) + ylab('% of Differentially Expressed Genes') +  xlab('Fold Change') +
         ggtitle('Effect of filtering thresholds in SFARI Genes') + theme_minimal())

```

<br>

***

## Grouping Genes by SFARI Gene Score

<br>

### Gene Expression
<br>

<br><br>

#### Normalised data

- **The higher the SFARI score, the higher the mean expression of the gene:** This pattern is quite strong and it doesn't have any biological interpretation, so it's probably bias in the SFARI score assignment

- **The higher the SFARI score, the lower the standard deviation:** This pattern is not as strong, but it is weird because the data was originally heteroscedastic with a positive relation between mean and variance, so the fact that the relation now seems to have reversed could mean that the vst normalisation ended up affecting the highly expressed genes more than it should have when trying to correct their higher variance

```{r, warning=FALSE, fig.width=10}
plot_data = data.frame('ID'=rownames(datExpr), 'MeanExpr'=rowMeans(datExpr), 'SDExpr'=apply(datExpr,1,sd)) %>% 
            left_join(DE_info, by='ID')

p1 = ggplotly(plot_data %>% ggplot(aes(gene.score, MeanExpr, fill=gene.score)) + geom_boxplot() + 
              scale_fill_manual(values=SFARI_colour_hue(r=c(1:6,8,7))) + theme_minimal() +
              theme(legend.position='none'))

p2 = ggplotly(plot_data %>% ggplot(aes(gene.score, SDExpr, fill=gene.score)) + geom_boxplot() + 
              scale_fill_manual(values=SFARI_colour_hue(r=c(1:6,8,7))) + theme_minimal() +
              ggtitle('Mean Expression (left) and SD (right) by SFARI score') + 
              theme(legend.position='none'))

subplot(p1, p2, nrows=1)

rm(plot_data, p1, p2)
```
<br>

#### Raw data

Just to corroborate that the relation between sd and SFARI score used to be in the opposite direction before the normalisation: **The higher the SFARI score the higher the mean expression and the higher the standard deviation**

*There are a lot of outliers, but the plot is interactive so you can zoom in

```{r, warning=FALSE, fig.width=10}

# Save preprocessed results
datExpr_prep = datExpr
datMeta_prep = datMeta
DE_info_prep = DE_info

load('./../Data/filtered_raw_data.RData')

plot_data = data.frame('ID'=rownames(datExpr), 'MeanExpr'=rowMeans(datExpr), 'SDExpr'=apply(datExpr,1,sd)) %>% 
            left_join(DE_info, by='ID')

p1 = ggplotly(plot_data %>% ggplot(aes(gene.score, MeanExpr, fill=gene.score)) + geom_boxplot() + 
              scale_fill_manual(values=SFARI_colour_hue(r=c(1:6,8,7))) + theme_minimal() +
              theme(legend.position='none'))

p2 = ggplotly(plot_data %>% ggplot(aes(gene.score, SDExpr, fill=gene.score)) + geom_boxplot() + 
              scale_fill_manual(values=SFARI_colour_hue(r=c(1:6,8,7))) + theme_minimal() +
              ggtitle('Mean Expression (left) and SD (right) by SFARI score') + 
              theme(legend.position='none'))

subplot(p1, p2, nrows=1)

rm(plot_data, p1, p2)

```

Return to normalised version of the data
```{r}
# Save preprocessed results
datExpr = datExpr_prep
datMeta = datMeta_prep
DE_info = DE_info_prep

rm(datExpr_prep, datMeta_prep, DE_info_prep)
```
<br>

### Log Fold Change

**There seems to be a negative relation between SFARI score and log fold change** when it would be expected to be either positively correlated or independent from each other (this last one because there are other factors that determine if a gene is releated to Autism apart from differences in gene expression)
<br><br>

[Wikipedia](https://en.wikipedia.org/wiki/Fold_change) mentions the likely explanation for this: **"A disadvantage and serious risk of using fold change in this setting is that it is biased and may misclassify differentially expressed genes with large differences (B − A) but small ratios (B/A), leading to poor identification of changes at high expression levels"**.
<br><br>

**Based on this, since we saw there is a strong relation between SFARI score and mean expression, the bias in log fold change affects mainly genes with high SFARI scores, which would be the ones we are most interested in.**
<br><br>

On top of this, I believe this effect is made more extreme by the pattern found in the previous plots, since the higher expressed genes were the most affected by the normalisation transformation, ending up with a smaller variance than the rest of the data, which is related to smaller ratios. (This is a constant problem independently of the normalisation function used).

```{r}
ggplotly(DE_info %>% ggplot(aes(x=gene.score, y=abs(log2FoldChange), fill=gene.score)) + 
         geom_boxplot() + scale_fill_manual(values=SFARI_colour_hue(r=c(1:6,8,7))) + xlab('SFARI Gene scores') + ylab('log Fold Change Magnitude') +
         theme_minimal() + theme(legend.position='none'))
```
<br>

### Effects of modifying filtering threshold by SFARI score

**The higher the SFARI score, the higher the percentage of genes that get filtered by differential expression.**

- This pattern is generally consistent on all log fold change thresholds

- All SFARI scores except 6 are more affected by the filtering than the genes with Neuronal-related functional annotations

- SFARI gene groups 1 and 2 are more affected than the average gene, although the number of genes with SFARI score 1 is small, so this result is not very robust

- At a threshold of log2(1.07), we lose all genes belonging to score 1 and 2

- Using the null hypothesis $H_0: lfc=0$, 273/903 SFARI genes are statistically significant (30%)

```{r lfc_by_SFARI_score_plot, fig.width=10, warning=FALSE, message=FALSE}
lfc_list = seq(1, 1.2, 0.01)

all_counts = data.frame('group'='All', 'n'=as.character(nrow(DE_info)))
Neuronal_counts = data.frame('group'='Neuronal', n=as.character(sum(DE_info$Neuronal)))
lfc_counts_all = DE_info %>% group_by(`gene-score`) %>% tally %>%
                 mutate('group'=as.factor(`gene-score`), 'n'=as.character(n)) %>%
                 dplyr::select(group, n) %>%
                 bind_rows(Neuronal_counts, all_counts) %>%
                 mutate('lfc'=-1) %>%  dplyr::select(lfc, group, n)

for(lfc in lfc_list){
  
  # Recalculate DE_info with the new threshold (p-values change)
  DE_genes = results(dds, lfcThreshold=log2(lfc), altHypothesis='greaterAbs') %>% data.frame
  
  DE_genes = DE_genes %>% mutate('ID'=rownames(.)) %>% left_join(SFARI_genes, by='ID') %>% 
             mutate(`gene-score`=ifelse(is.na(`gene-score`), 'Others', `gene-score`)) %>%
             distinct(ID, .keep_all = TRUE) %>% left_join(GO_neuronal, by='ID') %>%
             mutate(Neuronal=ifelse(is.na(Neuronal), 0, Neuronal)) %>%
             mutate(gene.score=ifelse(`gene-score`=='Others' & Neuronal==1, 'Neuronal', `gene-score`))
  
  DE_genes = DE_genes %>% filter(padj<0.05 & abs(log2FoldChange)>log2(lfc))

  
  # Calculate counts by groups
  all_counts = data.frame('group'='All', 'n'=as.character(nrow(DE_genes)))
  Others_counts = data.frame('group'='Others', n=as.character(sum(DE_genes$gene.score == 'Others')))
  Neuronal_counts = data.frame('group'='Neuronal', n=as.character(sum(DE_genes$Neuronal)))
  lfc_counts = DE_genes %>% group_by(`gene-score`) %>% tally %>%
               mutate('group'=`gene-score`, 'n'=as.character(n)) %>%
               bind_rows(Neuronal_counts, all_counts) %>%
               mutate('lfc'=lfc) %>% dplyr::select(lfc, group, n)
  
  
  # Update lfc_counts_all
  lfc_counts_all = lfc_counts_all %>% bind_rows(lfc_counts)
}

# Add missing entries with 0s
lfc_counts_all = expand.grid('group'=unique(lfc_counts_all$group), 'lfc'=unique(lfc_counts_all$lfc)) %>% 
  left_join(lfc_counts_all, by=c('group','lfc')) %>% replace(is.na(.), 0)

# Calculate percentage of each group remaining
tot_counts = DE_info %>% group_by(`gene-score`) %>% tally() %>% filter(`gene-score`!='Others') %>%
             mutate('group'=`gene-score`, 'tot'=n) %>% dplyr::select(group, tot) %>%
             bind_rows(data.frame('group'='Neuronal', 'tot'=sum(DE_info$Neuronal)),
                       data.frame('group' = 'Others', 'tot' = sum(DE_info$gene.score == 'Others')),
                       data.frame('group'='All', 'tot'=nrow(DE_info)))

lfc_counts_all = lfc_counts_all %>% filter(lfc!=-1) %>% #, group!='Others') %>% 
                 left_join(tot_counts, by='group') %>% mutate('perc'=round(100*as.numeric(n)/tot,2))


# Plot change of number of genes
ggplotly(lfc_counts_all %>% filter(group != 'All') %>% ggplot(aes(lfc, perc, color=group)) + geom_point(aes(id=n)) + geom_line() + 
         scale_color_manual(values=SFARI_colour_hue(r=c(1:6,8,7))) + ylab('% of Differentially Expressed Genes') +  xlab('Fold Change') + 
         ggtitle('Effect of filtering thresholds by SFARI score') + theme_minimal())

rm(lfc_list, all_counts, Neuronal_counts, lfc_counts_all, lfc, lfc_counts, lfc_counts_all, tot_counts, lfc_counts_all, Others_counts)
```

---

#### Session info

```{r}
sessionInfo()
```
<br><br>

